packages/next/build/index.ts

```diff
@@ -7,24 +7,76 @@ import {generateBuildId} from './generate-build-id'
 import {writeBuildId} from './write-build-id'
 import {isWriteable} from './is-writeable'
 import {runCompiler, CompilerResult} from './compiler'
+import globModule from 'glob'
+import {promisify} from 'util'
+import {stringify} from 'querystring'
 
-export default async function build (dir: string, conf = null, lambdas: boolean = false): Promise<void> {
+const glob = promisify(globModule)
+
+function collectPages (directory: string, pageExtensions: string[]): Promise<string[]> {
+  return glob(`**/*.+(${pageExtensions.join('|')})`, {cwd: directory})
+}
+
+export default async function build (dir: string, conf = null): Promise<void> {
   if (!await isWriteable(dir)) {
     throw new Error('> Build directory is not writeable. https://err.sh/zeit/next.js/build-dir-not-writeable')
   }
 
   const config = loadConfig(PHASE_PRODUCTION_BUILD, dir, conf)
-  const lambdasOption = config.lambdas ? config.lambdas : lambdas
-  const distDir = join(dir, config.distDir)
   const buildId = await generateBuildId(config.generateBuildId, nanoid)
+  const distDir = join(dir, config.distDir)
+  const pagesDir = join(dir, 'pages')
+
+  const pagePaths = await collectPages(pagesDir, config.pageExtensions)
+  type Result = {[page: string]: string}
+  const pages: Result = pagePaths.reduce((result: Result, pagePath): Result => {
+    let page = `/${pagePath.replace(new RegExp(`\\.+(${config.pageExtensions.join('|')})$`), '').replace(/\\/g, '/')}`.replace(/\/index$/, '')
+    page = page === '' ? '/' : page
+    result[page] = pagePath
+    return result
+  }, {})
+
+  let entrypoints
+  if (config.target === 'serverless') {
+    const serverlessEntrypoints: any = {}
+    // Because on Windows absolute paths in the generated code can break because of numbers, eg 1 in the path,
+    // we have to use a private alias
+    const pagesDirAlias = 'private-next-pages'
+    const dotNextDirAlias = 'private-dot-next'
+    const absoluteAppPath = pages['/_app'] ? join(pagesDirAlias, pages['/_app']).replace(/\\/g, '/') : 'next/dist/pages/_app'
+    const absoluteDocumentPath = pages['/_document'] ? join(pagesDirAlias, pages['/_document']).replace(/\\/g, '/') : 'next/dist/pages/_document'
+    const absoluteErrorPath = pages['/_error'] ? join(pagesDirAlias, pages['/_error']).replace(/\\/g, '/') : 'next/dist/pages/_error'
+    Object.keys(pages).forEach(async (page) => {
+      if (page === '/_app' || page === '/_document') {
+        return
+      }
+
+      const absolutePagePath = join(pagesDirAlias, pages[page]).replace(/\\/g, '/')
+      const bundleFile = page === '/' ? '/index.js' : `${page}.js`
+      const query = stringify({
+        page,
+        absolutePagePath,
+        absoluteAppPath,
+        absoluteDocumentPath,
+        absoluteErrorPath,
+        distDir: dotNextDirAlias,
+        buildId,
+        assetPrefix: config.assetPrefix,
+        generateEtags: config.generateEtags
+      })
```

Comment:
I think we might want some type checking from the serverless loader here

Reply:
Good catch 👍 Added!